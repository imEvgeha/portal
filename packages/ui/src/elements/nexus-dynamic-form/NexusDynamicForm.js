import React, {Fragment, useCallback, useContext, useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {default as AKForm, ErrorMessage} from '@atlaskit/form';
import {NexusModalContext} from '@vubiquity-nexus/portal-ui/lib/elements/nexus-modal/NexusModal';
import classnames from 'classnames';
import {get, isEmpty, mergeWith, set} from 'lodash';
import moment from 'moment';
import {useDispatch} from 'react-redux';
import {Link, useParams} from 'react-router-dom';
import PropagateForm from '../../../../../src/pages/title-metadata/components/title-metadata-details/components/PropagateForm';
import {clearTitleMetadataFilter} from '../../../../../src/pages/title-metadata/titleMetadataActions';
import PropagateButton from '../nexus-person/elements/PropagateButton/PropagateButton';
import ButtonsBuilder from './components/ButtonsBuilder/ButtonsBuilder';
import {buildSection, getAllFields, getProperValue, getProperValues} from './utils';
import {CAST_AND_CREW_TITLE, CORE_TITLE_SECTION, EPISODE, PROPAGATE_TITLE, SEASON, SERIES, VIEWS} from './constants';
import './NexusDynamicForm.scss';

const NexusDynamicForm = ({
    schema,
    initialData,
    onSubmit,
    canEdit,
    selectValues,
    isSaving,
    containerRef,
    isTitlePage,
    searchPerson,
    generateMsvIds,
    regenerateAutoDecoratedMetadata,
    hasButtons,
    setRefresh,
    castCrewConfig,
    seasonPersons,
}) => {
    const dispatch = useDispatch();

    const {openModal, closeModal} = useContext(NexusModalContext);
    const [disableSubmit, setDisableSubmit] = useState(true);
    const [update, setUpdate] = useState(false);
    const [validationErrorCount, setValidationErrorCount] = useState(0);
    const routeParams = useParams();

    const view = canEdit ? VIEWS.EDIT : VIEWS.VIEW;

    const {fields} = schema;

    useEffect(() => {
        update && setUpdate(false);
    }, [update]);

    useEffect(() => {
        // eslint-disable-next-line prefer-destructuring
        const firstErrorElement = document.getElementsByClassName('nexus-c-field--error')[0];
        if (firstErrorElement) firstErrorElement.scrollIntoView(false);
    }, [validationErrorCount]);

    const onCancel = () => {
        setRefresh(prev => !prev);
        setUpdate(true);
        setValidationErrorCount(0);
    };

    const showValidationError = () => {
        const errorsCount = document.getElementsByClassName('nexus-c-field--error').length;
        setValidationErrorCount(errorsCount);
    };

    const buttonsBuilder = (dirty, reset, errors) => {
        return (
            <>
                {errors > 0 && (
                    <div className="nexus-c-dynamic-form__validation-msg">
                        <ErrorMessage>
                            {errors} {errors === 1 ? 'error' : 'errors'} on page
                        </ErrorMessage>
                    </div>
                )}
                <ButtonsBuilder
                    dirty={dirty}
                    reset={reset}
                    validationErrorCount={validationErrorCount}
                    errors={validationErrorCount}
                    disableSubmit={disableSubmit}
                    canEdit={canEdit}
                    isSaving={isSaving}
                    isEmpty={isEmpty}
                    onCancel={onCancel}
                    seasonPersons={seasonPersons}
                    showValidationError={showValidationError()}
                />
            </>
        );
    };

    const validDateRange = values => {
        let areValid = true;
        const allFields = getAllFields(fields);
        Object.keys(allFields)
            .filter(key => allFields[key].type === 'dateRange')
            .forEach(key => {
                const {startDate, endDate} = values[key];
                if (moment(startDate).isAfter(endDate) || moment(endDate).isBefore(startDate)) {
                    areValid = false;
                }
            });
        return areValid;
    };

    const handleOnSubmit = (values, initialData) => {
        setValidationErrorCount(0);
        if (validDateRange(values)) {
            let correctValues = {};
            const allValues = getAllFields(fields);
            const properValues = getProperValues(fields, values);
            Object.values(allValues).forEach(({type, path}) => {
                const defaultValue = getProperValue(type, null, path, fields);
                correctValues = {...(type === 'array' && defaultValue), ...correctValues};
            });
            Object.keys(properValues).forEach(key => set(correctValues, key, properValues[key]));

            const valuesData = mergeWith({}, initialData, correctValues, (obj, src) => {
                if (Array.isArray(src)) {
                    return src;
                }
                // keep original null value if updated value is object and all its properties are falsy
                // non object values are null already if not edited
                else if (obj === null && typeof src === 'object') {
                    if (!src) return null;
                    if (
                        !Object.keys(src).some(k => {
                            if (Array.isArray(src[k]))
                                // if value is array
                                return src[k].length;
                            else if (typeof src[k] === 'object' && src[k] !== null)
                                // if value is object
                                return Object.keys(src[k]).length;
                            return src[k]; // else return value
                        })
                    )
                        return null;
                }
            });

            onSubmit(valuesData, initialData);
        }
    };

    const createLink = contentType => {
        const baseUrl = '/metadata/?parentId=';
        const id = get(initialData, 'id', '');
        return `/${routeParams.realm}${baseUrl}${id}&contentType=${contentType === SERIES ? SEASON : EPISODE}`;
    };

    const showAll = () => {
        if (isTitlePage) {
            const allowedContents = [SEASON, SERIES];
            const contentType = get(initialData, 'contentType', '');
            if (allowedContents.includes(contentType)) {
                return (
                    <div className="nexus-c-dynamic-form__show-all">
                        <Link onClick={() => dispatch(clearTitleMetadataFilter())} to={createLink(contentType)}>
                            Show all {contentType === SERIES ? 'seasons' : 'episodes'}
                        </Link>
                    </div>
                );
            }
        }
        return null;
    };

    const closePropagateModal = () => {
        closeModal();
        setUpdate(prev => !prev);
    };

    const showPropagateModal = useCallback((getValues, setFieldValue) => {
        openModal(<PropagateForm getValues={getValues} setFieldValue={setFieldValue} onClose={closePropagateModal} />, {
            title: PROPAGATE_TITLE,
            width: 'small',
        });
    }, []);

    return (
        <div className="nexus-c-dynamic-form">
            <AKForm onSubmit={values => handleOnSubmit(values, initialData)}>
                {({formProps, dirty, reset, getValues, setFieldValue}) => (
                    <form {...formProps}>
                        {hasButtons && buttonsBuilder(dirty, reset, validationErrorCount)}
                        <div
                            ref={containerRef}
                            className={classnames('nexus-c-dynamic-form__tab-container', {
                                'nexus-c-dynamic-form__tab-container--title': isTitlePage,
                            })}
                        />
                        <div
                            className={classnames('nexus-c-dynamic-form__tab-content', {
                                'nexus-c-dynamic-form__tab-content--title': isTitlePage,
                            })}
                        >
                            {fields.map(({title = '', sections = []}, index) => (
                                <div
                                    key={`tab-${title}`}
                                    id={`tab-${index}`}
                                    className="nexus-c-dynamic-form__section-start"
                                >
                                    {sections.map(
                                        ({
                                            title: sectionTitle = '',
                                            fields = {},
                                            isGridLayout = false,
                                            tabs,
                                            subTabs,
                                            prefix,
                                        }) => (
                                            <Fragment key={`section-${sectionTitle}`}>
                                                <h3 className="nexus-c-dynamic-form__section-title">
                                                    {sectionTitle === CAST_AND_CREW_TITLE && isTitlePage && canEdit ? (
                                                        <div className="nexus-c-dynamic-form__additional-option">
                                                            {sectionTitle}
                                                            <PropagateButton
                                                                onClick={() =>
                                                                    showPropagateModal(getValues, setFieldValue)
                                                                }
                                                            />
                                                        </div>
                                                    ) : (
                                                        sectionTitle
                                                    )}
                                                </h3>
                                                {sectionTitle === CORE_TITLE_SECTION && showAll()}
                                                {buildSection(
                                                    fields,
                                                    getValues,
                                                    view,
                                                    generateMsvIds,
                                                    regenerateAutoDecoratedMetadata,
                                                    setRefresh,
                                                    {
                                                        selectValues,
                                                        initialData,
                                                        setFieldValue,
                                                        update,
                                                        config: schema.config || [],
                                                        isGridLayout,
                                                        searchPerson,
                                                        castCrewConfig,
                                                        tabs,
                                                        subTabs,
                                                        setDisableSubmit,
                                                        prefix,
                                                        isTitlePage,
                                                        setUpdate,
                                                    }
                                                )}
                                            </Fragment>
                                        )
                                    )}
                                </div>
                            ))}
                        </div>
                    </form>
                )}
            </AKForm>
        </div>
    );
};

NexusDynamicForm.propTypes = {
    schema: PropTypes.object,
    initialData: PropTypes.object,
    onSubmit: PropTypes.func,
    canEdit: PropTypes.bool,
    selectValues: PropTypes.object,
    containerRef: PropTypes.any,
    isTitlePage: PropTypes.bool,
    searchPerson: PropTypes.func,
    generateMsvIds: PropTypes.func,
    isSaving: PropTypes.bool,
    regenerateAutoDecoratedMetadata: PropTypes.func,
    hasButtons: PropTypes.bool,
    setRefresh: PropTypes.func,
    castCrewConfig: PropTypes.object,
    storedInitialData: PropTypes.object,
    seasonPersons: PropTypes.array,
};

NexusDynamicForm.defaultProps = {
    schema: {},
    initialData: {},
    onSubmit: undefined,
    canEdit: false,
    selectValues: {},
    containerRef: null,
    isTitlePage: false,
    searchPerson: undefined,
    generateMsvIds: undefined,
    isSaving: false,
    regenerateAutoDecoratedMetadata: undefined,
    hasButtons: true,
    setRefresh: () => null,
    castCrewConfig: {},
    storedInitialData: null,
    seasonPersons: [],
};

export default NexusDynamicForm;
