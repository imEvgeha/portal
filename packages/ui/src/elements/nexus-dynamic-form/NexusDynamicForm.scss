@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/mixins';
@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/colors';

// only nest modifier within block, not elements
.nexus-c-dynamic-form {
    width: 80%;
    margin: auto;
}

.nexus-c-dynamic-form__tab-container {
    width: calc((100% - 64px) * 0.8);
    margin: auto;
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    position: fixed;
    top: 131px;
    transition: top linear 0.5s;
    background-color: map-get($neutrals, N0);
    z-index: 1;
}

.nexus-c-dynamic-form__tab-container--title {
    top: 175px;
}

.nexus-c-dynamic-form__tab-container--shrinked-title {
    top: 82px;
}

.nexus-c-dynamic-form__tab {
    display: flex;
    justify-content: center;
    @include padding(1, 6);
    border-bottom: 1px solid map-get($neutrals, N40);
    cursor: pointer;
    transition: color 0.2s ease-in;

    &:hover {
        color: map-get($blues, B400);
    }

    &--is-active {
        color: map-get($blues, B400);
        border-bottom: 2px solid map-get($blues, B400);
        cursor: default;
    }
}

.nexus-c-dynamic-form__section-title {
    @include padding(13, 0, 2);
    font-weight: bold;
    border-bottom: 2px solid map-get($neutrals, N40);
}

.nexus-c-dynamic-form__actions-container {
    display: flex;
    align-items: center;
    position: absolute;
    bottom: 10px;
    right: 10%;
    z-index: 2;
}
.nexus-c-dynamic-form__discard-button,
.nexus-c-dynamic-form__submit-button {
    border-radius: 8px;
    border: 1px solid lightgray;
}

.nexus-c-dynamic-form__status {
    width: 30px;
}

.success {
    background-color: map-get($greens, G14);
}

.updated {
    background-color: map-get($yellows, Y14);
}

.error {
    background-color: map-get($reds, R14);
}

.nexus-c-dynamic-form__validation-msg {
    position: absolute;
    top: 24px;
    z-index: 1;
    font-size: 1.3rem;
    right: 245px;
}

.nexus-c-dynamic-form__title-validation-msg {
    position: absolute;
    top: 7%;
    z-index: 1;
    font-size: 1.3rem;
    right: 35%;
}

.nexus-c-dynamic-form__field {
    .nexus-c-field {
        > div {
            display: flex;
            align-items: center;
        }

        .nexus-c-field__label {
            width: 25%;
            color: map-get($neutrals, N300);
            &--grid {
                width: 50%;
            }
        }
    }
}

.nexus-c-dynamic-form__section--grid {
    display: grid;
    grid-template-columns: 50% 50%;
}

.nexus-c-field__value-section {
    width: 100%;
}

.nexus-c-dynamic-form__section-start {
    scroll-margin: 80px;
}
