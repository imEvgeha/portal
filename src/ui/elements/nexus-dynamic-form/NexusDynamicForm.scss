@import '../../../styles/atlaskit/mixins';
@import '../../../styles/atlaskit/colors';

.nexus-c-dynamic-form {
    width: 80%;
    margin: auto;
}

.nexus-c-dynamic-form__tab-container {
    width: calc((100% - 64px) * 0.8);
    margin: auto;
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    border-bottom: 1px solid map-get($neutrals, N40);
    position: fixed;
    top: 131px;
    transition: top linear 0.5s;
    background-color: map-get($neutrals, N0);
    z-index: 1;
}

.nexus-c-dynamic-form__tab-container--title {
    top: 175px;
}

.nexus-c-dynamic-form__tab-container--shrinked-title {
    top: 82px;
}

.nexus-c-dynamic-form__tab-content {
    @include padding(32, 0, 0, 0);
}

.nexus-c-dynamic-form__tab-content--title {
    @include padding(46, 0, 0, 0);
}

.nexus-c-dynamic-form__tab {
    display: flex;
    justify-content: center;
    @include padding(1, 6);
    border-bottom: 1px solid map-get($neutrals, N40);
    cursor: pointer;
    transition: color 0.2s ease-in;

    &:hover {
        color: map-get($blues, B400);
    }

    &--is-active {
        color: map-get($blues, B400);
        border-bottom: 2px solid map-get($blues, B400);
        cursor: default;
    }
}

.nexus-c-dynamic-form__section-title {
    @include padding(13, 0, 2);
    font-weight: bold;
    border-bottom: 2px solid map-get($neutrals, N40);
}

.nexus-c-dynamic-form__cancel-button,
.nexus-c-dynamic-form__submit-button,
.nexus-c-dynamic-form__edit-button {
    position: absolute;
    top: 24px;
    right: 24px;
    z-index: 1;
}

.nexus-c-dynamic-form__submit-button {
    right: 100px;
}

.nexus-c-dynamic-form__cancel-button--title,
.nexus-c-dynamic-form__submit-button--title,
.nexus-c-dynamic-form__edit-button--title {
    position: absolute;
    top: 24px;
    right: 31.5%;
    z-index: 1;
}

.nexus-c-dynamic-form__submit-button--title {
    right: calc(31.5% + 80px);
}

.nexus-c-dynamic-form__field {
    .nexus-c-field {
        > div {
            display: flex;
            align-items: center;
        }

        .nexus-c-field__label {
            width: 25%;
            color: map-get($neutrals, N300);
        }
    }
}

.nexus-c-field__value-section {
    width: 100%;
}
