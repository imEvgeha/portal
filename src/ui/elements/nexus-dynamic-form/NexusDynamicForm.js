import React, {Fragment, useState} from 'react';
import PropTypes from 'prop-types';
import Button from '@atlaskit/button';
import {default as AKForm} from '@atlaskit/form';
import SectionTab from './components/SectionTab/SectionTab';
import {buildSection, getProperValues} from './utils';
import {VIEWS} from './constants';
import './NexusDynamicForm.scss';

const NexusDynamicForm = ({schema = [], initialData, onSubmit, isEdit, selectValues, containerRef}) => {
    const tabs = schema.map(({title = ''}) => title);
    const [selectedTab, setSelectedTab] = useState(tabs[0]);
    const [view, setView] = useState(isEdit ? VIEWS.VIEW : VIEWS.CREATE);

    const buildTabs = () => {
        return tabs.map(tab => (
            <SectionTab key={tab} section={tab} onClick={() => setSelectedTab(tab)} isActive={selectedTab === tab} />
        ));
    };

    const buildButtons = (dirty, submitting, reset) => {
        return view !== VIEWS.VIEW ? (
            <>
                <Button
                    type="submit"
                    className="nexus-c-dynamic-form__submit-button"
                    appearance="primary"
                    isDisabled={!dirty || submitting}
                >
                    Save changes
                </Button>
                <Button
                    className="nexus-c-dynamic-form__cancel-button"
                    onClick={() => {
                        reset();
                        setView(VIEWS.VIEW);
                    }}
                >
                    Cancel
                </Button>
            </>
        ) : (
            <Button
                className="nexus-c-dynamic-form__edit-button"
                appearance="primary"
                onClick={() => setView(VIEWS.EDIT)}
            >
                Edit
            </Button>
        );
    };

    const handleOnSubmit = values => {
        setView(VIEWS.VIEW);
        const properValues = getProperValues(schema, values);
        onSubmit(properValues);
    };

    return (
        <div className="nexus-c-dynamic-form">
            <AKForm onSubmit={values => handleOnSubmit(values)}>
                {({formProps, dirty, submitting, reset, getValues, setFieldValue}) => (
                    <form {...formProps}>
                        {buildButtons(dirty, submitting, reset)}
                        <div ref={containerRef} className="nexus-c-dynamic-form__tab-container">
                            {buildTabs()}
                        </div>
                        <div className="nexus-c-dynamic-form__tab-content">
                            {schema.map(({title = '', sections = []}) => (
                                <Fragment key={`tab-${title}`}>
                                    {sections.map(({title: sectionTitle = '', fields = {}}) => (
                                        <Fragment key={`section-${sectionTitle}`}>
                                            <h3 id={sectionTitle} className="nexus-c-dynamic-form__section-title">
                                                {sectionTitle}
                                            </h3>
                                            {buildSection(fields, getValues, view, {
                                                selectValues,
                                                initialData,
                                                setFieldValue,
                                            })}
                                        </Fragment>
                                    ))}
                                </Fragment>
                            ))}
                        </div>
                    </form>
                )}
            </AKForm>
        </div>
    );
};

NexusDynamicForm.propTypes = {
    schema: PropTypes.array.isRequired,
    initialData: PropTypes.object,
    onSubmit: PropTypes.func,
    isEdit: PropTypes.bool,
    selectValues: PropTypes.object,
    containerRef: PropTypes.any,
};

NexusDynamicForm.defaultProps = {
    initialData: {},
    onSubmit: undefined,
    isEdit: false,
    selectValues: {},
    containerRef: null,
};

export default NexusDynamicForm;
