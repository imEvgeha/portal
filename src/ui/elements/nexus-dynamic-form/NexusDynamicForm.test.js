import React from 'react';
import {shallow} from 'enzyme';
import NexusDynamicForm from './NexusDynamicForm';

describe('NexusDynamicForm', () => {
    let wrapper = null;
    const schema = [
        {
            title: 'Amazing Tab',
            sections: [
                {
                    title: 'Best section 2020',
                    fields: [
                        {
                            label: 'Awesome field',
                            type: 'text',
                        },
                    ],
                },
            ],
        },
    ];

    beforeEach(() => {
        wrapper = shallow(<NexusDynamicForm schema={schema}/>);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render tab container and all tabs within', () => {
        const tabs = schema.map(({title = ''}) => title);
        const tabContainer = wrapper.find('.nexus-c-dynamic-form__tab-container');
        expect(tabContainer.length).toEqual(1); // Container exists

        expect(tabContainer.children().length).toEqual(tabs.length); // Container has all tabs

        tabContainer.children().forEach((child, index) => {
            expect(child.dive().text()).toEqual(tabs[index]); // All tabs are named properly
        })
    });
});
