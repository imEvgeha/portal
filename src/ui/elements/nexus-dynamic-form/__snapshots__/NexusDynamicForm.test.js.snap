// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusDynamicForm should match snapshot 1`] = `
<div
  className="nexus-c-dynamic-form"
>
  <div
    className="nexus-c-dynamic-form__tab-container"
  >
    <SectionTab
      isActive={true}
      key="Amazing Tab"
      onClick={[Function]}
      section="Amazing Tab"
    />
  </div>
  <div
    className="nexus-c-dynamic-form__tab-content"
  >
    <h3
      className="nexus-c-dynamic-form__section-title"
      id="Best section 2020"
    >
      Best section 2020
    </h3>
    <div
      key="field-Awesome field"
    >
      Awesome field
    </div>
  </div>
</div>
`;
